### PR TITLE
Add UniformBinning to schemav2

### PR DESCRIPTION
**Pull request by eguiraud**
When defining a Binning, the `edges` attribute can now be either a list of floats (like before) or an instance of UniformBinning(n, low, high). When using the latter option, the Binning correction is optimized to use a simple arithmetic formula rather than binary search to look up the bin index.

This is a fresh (better) take on #181 .

Example usage:

```python
from correctionlib import schemav2 as schema

if __name__ == "__main__":
    corr = schema.Correction(
        name="test",
        version=2,
        inputs=[schema.Variable(name="x", type="real")],
        output=schema.Variable(name="a scale", type="real"),
        data=schema.Binning(
            nodetype="binning",
            input="x",
            edges=schema.UniformBinning(n=2, low=0., high=2.),
            content=[1.0, 2.0],
            flow="error",
        ),
    )
    e = corr.to_evaluator()
    print(e.evaluate(0.5))
    print(e.evaluate(1.0))
    print(e.evaluate(3.0)) # errors out
```

To do:
- [x] add tests
- [x] add support for MultiBinning

### CODE REVIEW

File: src/correction.cc [hunk 1]
```diff
@@ -319,17 +319,35 @@ double HashPRNG::evaluate(const std::vector<Variable::Type>& values) const {
...
+      double val = edge.GetDouble();
+      if ( edges.size() > 0 && edges.back() >= val ) { throw std::runtime_error("binning edges are not monotone increasing"); }
+      edges.push_back(val);
+    }
+    if ( edges.size() != content.Size() + 1 ) {
+      throw std::runtime_error("Inconsistency in Binning: number of content nodes does not match binning");
+    }
+    _NonUniformBins bins{std::move(edges)};
+    bins_ = std::move(bins);
+  } else { // must be UniformBinning
```
Comment (nsmith-) [inline src/correction.cc:340]: `else if ( edgesObj.IsObject() )`
then a final else that throws an error

File: src/correction.cc [hunk 2]
```diff
@@ -319,17 +319,35 @@ double HashPRNG::evaluate(const std::vector<Variable::Type>& values) const {
...
+    const JSONObject uniformBins{edgesObj.GetObject()};
+    const auto n = uniformBins.getRequired<uint64_t>("n");
```
Comment (nsmith-) [inline src/correction.cc:342]: Check for n=0 here (since we don't always run the json through pydantic)

**Review by nsmith- — CHANGES_REQUESTED** [review]
Yeah this is the right approach I think.
Looks good, some small extra validation comments only

File: src/correction.cc (hunk 1 above)
Comment (eguiraud) [reply src/correction.cc]: done!

File: src/correction.cc (hunk 2 above)
Comment (eguiraud) [reply src/correction.cc:342]: done!

**Comment by eguiraud** [comment]
Addressed review comments, added some tests for Binning+UniformBinning, fixed a bug in an edge case that the test found.

I will work on adding support for MultiBinning as well as soon as possible (could also be a separate PR).

**Review by nsmith- — APPROVED** [review]
Looks great! Thanks so much for this addition, and sorry I didn't review it sooner
